interop-testing/src/main/java/io/grpc/testing/integration/AbstractInteropTest.java

```diff
@@ -1830,50 +1830,54 @@ private void assertServerStatsTrace(String method, Status.Code code,
       return;
     }
     AssertionError checkFailure = null;
-    boolean passed = false;
 
     if (metricsExpected()) {
       // Because the server doesn't restart between tests, it may still be processing the requests
       // from the previous tests when a new test starts, thus the test may see metrics from previous
       // tests.  The best we can do here is to exhaust all records and find one that matches the
       // given conditions.
+      boolean startRecordFound = false;
+      boolean endRecordFound = false;
       while (true) {
-        MetricsRecord serverStartRecord;
-        MetricsRecord serverEndRecord;
+        MetricsRecord serverRecord;
         try {
-          // On the server, the stats is finalized in ServerStreamListener.closed(), which can be
-          // run after the client receives the final status.  So we use a timeout.
-          serverStartRecord = serverStatsRecorder.pollRecord(5, TimeUnit.SECONDS);
-          serverEndRecord = serverStatsRecorder.pollRecord(5, TimeUnit.SECONDS);
+          serverRecord = serverStatsRecorder.pollRecord(5, TimeUnit.SECONDS);
         } catch (InterruptedException e) {
           throw new RuntimeException(e);
         }
-        if (serverEndRecord == null) {
+        if (serverRecord == null) {
           break;
         }
-        try {
-          checkStartTags(serverStartRecord, method);
-          checkEndTags(serverEndRecord, method, code);
-          if (requests != null && responses != null) {
-            checkCensus(serverEndRecord, true, requests, responses);
+        if (!startRecordFound) {
+          try {
+            checkStartTags(serverRecord, method);
```

Comment:
I don't think this approach is correct.  If its possible old tags could be in here, they shouldn't be validated, but discarded.  This is using exceptions as control flow, and it doesn't *look* right.

Reply:
This is a trial-and-error approach. This is not an usual approach, but it's the best I can find for this case.  Being trial-and-error implies involving error handling (in Java, it's the exceptions) in the control flow.

Reply:
This doesn't work.  It does the wrong thing.  And it's confusing.  There are a number of right ways to do it, so please don't treat this as the only one.  For example:

* If there could be data from multiple calls, then shutdown the server for each test method.   Stats are per server, right?  

* If that's impossible, then you need to collect all the record tags until there are no more, and then scan through to see if the method you are looking for exists.  `checkStartTags`  implies that `method` MUST be in there.   That is false in this case.

* Tags that WERE found are relevant because if we dont find it, we would like to know what was present.   Right now your code just keeps the exception of one of the tags, rather than all of them.  How are you going to debug that??

* The method name is not enough to find the correct tags.   We call the same method multiple times in the test, so finding the tags for a method may not actually correspond to the correct invocation.

* If you do consume multiple tags, you may accidentally prevent a subsequent test from passing.  If testA gets B's and A's tags, and then testB is not going to find its own.  

Please don't make this already complicated test class worse.